src/org/jitsi/impl/neomedia/transform/RetransmissionRequesterDelegate.java

```diff
@@ -0,0 +1,582 @@
+/*
+ * Copyright @ 2017 Atlassian Pty Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jitsi.impl.neomedia.transform;
+
+import java.io.*;
+import java.util.*;
+
+import org.jetbrains.annotations.*;
+import org.jitsi.impl.neomedia.*;
+import org.jitsi.impl.neomedia.rtcp.*;
+import org.jitsi.impl.neomedia.rtp.*;
+import org.jitsi.service.neomedia.*;
+import org.jitsi.service.neomedia.codec.*;
+import org.jitsi.service.neomedia.format.*;
+import org.jitsi.util.*;
+import org.jitsi.util.concurrent.*;
+
+/**
+ * Detects lost RTP packets for a particular <tt>RtpChannel</tt> and requests
+ * their retransmission by sending RTCP NACK packets.
+ *
+ * @author Boris Grozev
+ * @author George Politis
+ * @author bbaldino
+ */
+public class RetransmissionRequesterDelegate
+    extends SinglePacketTransformerAdapter
+    implements TransformEngine, RecurringRunnable
+{
+    /**
+     * If more than <tt>MAX_MISSING</tt> consecutive packets are lost, we will
+     * not request retransmissions for them, but reset our state instead.
+     */
+    public static final int MAX_MISSING = 100;
+
+    /**
+     * The maximum number of retransmission requests to be sent for a single
+     * RTP packet.
+     */
+    public static final int MAX_REQUESTS = 10;
+
+    /**
+     * The interval after which another retransmission request will be sent
+     * for a packet, unless it arrives. Ideally this should not be a constant,
+     * but should be based on the RTT to the endpoint.
+     */
+    public static final int RE_REQUEST_AFTER_MILLIS = 150;
+
+    /**
+     * The interval we'll ask the {@link RecurringRunnableExecutor} to check back
+     * in if there is no current work
+     * TODO(brian): i think we should actually be able to get rid of this and
+     * just rely on scheduled work and the 'work ready now' callback
+     */
+    public static final long WAKEUP_INTERVAL_MILLIS = 1000;
```

Comment:
Just a heads-up: there's currently no way to make the executor sleep indefinitely, and any value returned from getTimeUntilNextRun() that is <=0 triggers immediate execution.

Reply:
good to know...might be a nice future improvement for the executor.